synapse/storage/_base.py

```diff
@@ -569,12 +668,56 @@ def _simple_upsert_txn(self, txn, table, keyvalues, values, insertion_values={},
         sql = "INSERT INTO %s (%s) VALUES (%s)" % (
             table,
             ", ".join(k for k in allvalues),
-            ", ".join("?" for _ in allvalues)
+            ", ".join("?" for _ in allvalues),
         )
         txn.execute(sql, list(allvalues.values()))
         # successfully inserted
         return True
 
+    def _simple_upsert_txn_native_upsert(
+        self, txn, table, keyvalues, values, insertion_values={}, best_effort=False
+    ):
+        """
+        Use the native UPSERT functionality in recent PostgreSQL versions.
+        """
+        allvalues = {}
+        allvalues.update(keyvalues)
+        allvalues.update(values)
+        allvalues.update(insertion_values)
+
+        sql = (
+            "INSERT INTO %s (%s) VALUES (%s) "
+            "ON CONFLICT (%s) DO UPDATE SET %s "
+            "RETURNING (xmax = 0) AS inserted"
+        ) % (
+            table,
+            ", ".join(k for k in allvalues),
+            ", ".join("?" for _ in allvalues),
+            ", ".join(k for k in keyvalues),
+            ", ".join(k + "=EXCLUDED." + k for k in values),
+        )
+        try:
+            txn.execute(sql, list(allvalues.values()))
+        except self.database_engine.module.OperationalError as e:
+            # We only care about serialization errors, so check for it
+            if e.args[0] == "could not serialize access due to concurrent update":
```

Comment:
This is probably not safe, since I would be surprised if psycopg2 guaranteed that the error messages are stable. Please just use the error codes for this, as per: https://github.com/matrix-org/synapse/blob/b60749a1ec22269216a503ccd35aaf4454dc507d/synapse/storage/engines/postgres.py#L57-L63

Reply:
I was reading off https://www.postgresql.org/docs/10/transaction-iso.html#XACT-REPEATABLE-READ , that didn't give anything nice. That works better, though.